### ISSUE TITLE
get_all_licenses() : Syntax error with iLO 1 & iLO 2

### ISSUE THREAD

**Issue by Antalis** [seveas/python-hpilo#95]
Hi,
I want to get the license information for iLO1 and iLO2 (with last firmware), and I get an exception in both cases (with python-hpilo 3.3)  :

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/site-packages/hpilo.py", line 1149, in get_all_licenses
    return self._info_tag('RIB_INFO', 'GET_ALL_LICENSES', process=process)
  File "/usr/lib/python2.7/site-packages/hpilo.py", line 696, in _info_tag
    header, message = self._request(root)
  File "/usr/lib/python2.7/site-packages/hpilo.py", line 245, in _request
    message = self._parse_message(data)
  File "/usr/lib/python2.7/site-packages/hpilo.py", line 581, in _parse_message
    raise IloError(message, status)
hpilo.IloError: Syntax error: Line #0: syntax error near "" in the line: "". You may have tried to use a feature this iLO version or firmware version does not support.

**Comment by seveas** [comment]
get_all_licenses is only supported for ilo3 and ilo4. Unfortunately, the iLO's way of telling you this is saying you have a syntax error.

**Comment by Antalis** [comment]
Hi,
Ok with this method but is it possible to implement a second method ?
For example, this link https://bugzilla.mozilla.org/show_bug.cgi?id=873821  offers the following method : 
https://{$IP_ILO}/xmldata?item=CpqKey       and we have this result : 

iLO 1 : 
  <?xml version="1.0" ?> 
- <PROLIANTKEY>
  <KVER>1</KVER> 
  <LTYPE>FQL</LTYPE> 
  <IDATE>3-20-2012</IDATE> 
  <KEY>XXXXX-XXXXX-XXXXX-XXXXX-XXXXX</KEY> 
  <SN>ILOXXXXX</SN> 
  <SBSN>XXXXX</SBSN> 
  </PROLIANTKEY>

iLO 2 :
  <?xml version="1.0" ?> 
- <PROLIANTKEY>
  <KVER>1</KVER> 
  <LTYPE>FQL</LTYPE> 
  <LNAME>iLO 2 Advanced</LNAME> 
  <IDATE>7-15-2010</IDATE> 
  <KEY>XXXXX-XXXXX-XXXXX-XXXXX-XXXXX</KEY> 
  <SN>ILOUSEXXXXXX</SN> 
  <SBSN>USEXXXXXX</SBSN> 
  </PROLIANTKEY>

iLO 3 : 
  <?xml version="1.0" ?> 
- <PROLIANTKEY>
  <KVER>1</KVER> 
  <LTYPE>FQL</LTYPE> 
  <LNAME>iLO 3 Advanced</LNAME> 
  <IDATE>Mon Feb 2 12:13:59 2015</IDATE> 
  <KEY>XXXXX-XXXXX-XXXXX-XXXXX-XXXXX</KEY> 
  <SN>ILOCZXXXX</SN> 
  <SBSN>CZXXXXX</SBSN> 
  </PROLIANTKEY> 

iLO 4 : 
  <?xml version="1.0" ?> 
- <PROLIANTKEY>
  <KVER>1</KVER> 
  <LTYPE>FQL</LTYPE> 
  <LNAME>iLO 4 Advanced</LNAME> 
  <IDATE>Tue Mar 17 23:24:58 2015</IDATE> 
  <KEY>XXXXX-XXXXX-XXXXX-XXXXX-XXXXX</KEY> 
  <SN>ILOCZXXX</SN> 
  <SBSN>CZXXX</SBSN> 
  </PROLIANTKEY>

**Comment by seveas** [comment]
hpilo objects have an xmldata() object to get you that information

**Comment by Antalis** [comment]
Sorry but xmldata() does not indicate the license key,  and unfortunately that is the element I need.

**Comment by seveas** [comment]
Hmm. That would be a bug. Could you share debug information as outlined [here](http://seveas.github.io/python-hpilo/troubleshooting.html) please.

**Comment by Antalis** [comment]
I just send it to your email address.

**Comment by seveas** [comment]
I just noticed you used item=cpqKey, whereas python-hpilo uses item=all. I find it mildly irritating that HP decided that 'all' isn't really 'all'... Do you know of any other item= keys you can use?

**Comment by seveas** [comment]
As of 0ff3288 you can use xmldata(item='cpqkey') to get to the information you want.

**Comment by Antalis** [comment]
Sorry for my late return.
I tested this solution included in the latest update: 3.4, and i have this error : 
```python
Traceback (most recent call last):
  File "./test_ilo.py", line 79, in <module>
    Main()
  File "./test_ilo.py", line 55, in Main
    result = ilo.xmldata(item='cpqkey')
  File "/usr/lib/python2.7/site-packages/hpilo.py", line 1859, in xmldata
    return self._element_children_to_dict(etree.fromstring(data))
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1300, in XML
    parser.feed(text)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1642, in feed
    self._raiseerror(v)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1506, in _raiseerror
    raise err
xml.etree.ElementTree.ParseError: XML or text declaration not at start of entity: line 2, column 0
```

**Comment by seveas** [comment]
That is weird. Can you add a line saying `print(repr(data))` just before the call to `self._element_children_to_dict(etree.fromstring(data))` on line 1859 in hpilo.py and share the (redacted) output?

**Comment by Antalis** [comment]
In fact, with 'cpqkey', it begins with a first empty line ('\n') : 
```python
 ./test_ilo.py
'\n<?xml version="1.0"?><PROLIANTKEY><KVER>1</KVER><LTYPE>FQL</LTYPE><LNAME>iLO 2 Advanced</LNAME><IDATE>7-15-2010</IDATE><KEY>XXXXX-XXXXX-XXXXX-XXXXX-XXXXX</KEY><SN>ILOUSEXXXXXXX      </SN><SBSN>USEXXXXXXX      </SBSN></PROLIANTKEY>'
Traceback (most recent call last):
  File "./test_ilo.py", line 79, in <module>
    Main()
  File "./test_ilo.py", line 55, in Main
    result = ilo.xmldata(item='cpqkey')
  File "/usr/lib/python2.7/site-packages/hpilo.py", line 1860, in xmldata
    return self._element_children_to_dict(etree.fromstring(data))
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1300, in XML
    parser.feed(text)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1642, in feed
    self._raiseerror(v)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1506, in _raiseerror
    raise err
xml.etree.ElementTree.ParseError: XML or text declaration not at start of entity: line 2, column 0
```

**Comment by seveas** [comment]
ok, that's easy to fix/work around then: replace `etree.fromstring(data)` with `etree.fromstring(data.lstrip())` in hpilo.py line 1859. I'll commit that as a fix soon after some more testing.

**Comment by Antalis** [comment]
Thank you very much, it works :-)

```patch
- return self._element_children_to_dict(etree.fromstring(data))
+ return self._element_children_to_dict(etree.fromstring(data.lstrip()))
```
With iLO1, the result is a "dict" with this structure :
```python
{'idate': '3-20-2012',
 'key': 'XXXXX-XXXXX-XXXXX-XXXXX-XXXXX',
 'kver': 1,
 'ltype': 'FQL',
 'sbsn': 'YYYYYYYYYY',
 'sn': 'ILOYYYYYYYYYY'}
```
With iLO2 :
```python
{'idate': '7-15-2010',
 'key': 'XXXXX-XXXXX-XXXXX-XXXXX-XXXXX',
 'kver': 1,
 'lname': 'iLO 2 Advanced',
 'ltype': 'FQL',
 'sbsn': 'YYYYYYYYYY',
 'sn': 'ILOYYYYYYYYYY'}
```
With iLO3 :
```python
{'idate': 'Wed Aug 26 09:36:49 2015',
 'key': 'XXXXX-XXXXX-XXXXX-XXXXX-XXXXX',
 'kver': 1,
 'lname': 'iLO 3 Advanced',
 'ltype': 'FQL',
 'sbsn': 'YYYYYYYYYY',
 'sn': 'ILOYYYYYYYYYY'}
```
With iLO4 : 
```python
{'idate': 'Tue Mar 17 23:24:58 2015',
 'key': 'XXXXX-XXXXX-XXXXX-XXXXX-XXXXX',
 'kver': 1,
 'lname': 'iLO 4 Advanced',
 'ltype': 'FQL',
 'sbsn': 'YYYYYYYYYY',
 'sn': 'ILOYYYYYYYYYY'}
```

**Comment by seveas** [comment]
Awesome, thanks again for the report and all the testing!

**Comment by SLoeuillet** [comment]
xmldata fonction needs the very same fix in current git version (post 3.8)

        return self._element_children_to_dict(etree.fromstring(data.lstrip()))
